keps/sig-auth/3299-kms-v2-improvements/README.md

```diff
@@ -242,34 +243,79 @@ message StatusResponse {
+
+**Question: should writes to this resource be limited to the API server loopback identity?**
+
+**Question: should we have a status field with the Kubernetes version to prevent migration during differences in API server version?**
+
+```go
+// metadata.name is the ID of the API server
+type EncryptionState struct {
+    metav1.TypeMeta `json:",inline"`
+    // +optional
+    metav1.ObjectMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+
+    Spec EncryptionStateSpec `json:"spec" protobuf:"bytes,2,opt,name=spec"`
+
+    Status EncryptionStateStatus `json:"status" protobuf:"bytes,3,opt,name=status"`
+}
+
+// spec is empty
+type EncryptionStateSpec struct{}
+
+// status will be updated every 1 minute
+// KCM garbage collector will delete objects that have not been updated within the last 15 minutes
+type EncryptionStateStatus struct {
+	// running a storage migration when the API servers do not agree on the
+	// desired state will result in the etcd data being in an undefined state
+    EncryptionConfigurationHash Hash `json:"encryptionConfigurationHash" protobuf:"bytes,1,opt,name=encryptionConfigurationHash"`
+
+	// all resources that are encrypted are represented as individual items in this list
+	// note that this may get problamtic if we support */*
+    Resources []EncryptionStateResource `json:"resources" protobuf:"bytes,2,opt,name=resources"`
+
+	// the last time any status hash was updated
+	LastUpdated metav1.Time `json:"lastUpdated" protobuf:"bytes,3,opt,name=lastUpdated"`
+}
+
+type EncryptionStateResource struct {
+    Resource metav1.GroupResource `json:"resource" protobuf:"bytes,1,opt,name=resource"`
+
+	// an empty string from the hash value with a non-empty updated time means that
+	// a non-KMS v2 write key is in use.  This can be used with the `EncryptionConfigurationHash`
+	// to determine when the API servers are at the same state.
+    WriteKeyIDHash Hash `json:"writeKeyIDHash" protobuf:"bytes,2,opt,name=writeKeyIDHash"`
+
+	// note that this field does not cover KMS v1 and other legacy encryption modes that
+	// are set as read keys.  This is okay because those configurations are statically
+	// covered by the `EncryptionConfigurationHash`.  KMS v2 can dynamically change keys
+	// at runtime without any change to the `EncryptionConfigurationHash` and thus needs
+	// its own hash tracking mechanism.
+    ReadKeyIDHashes []Hash `json:"readKeyIDHashes" protobuf:"bytes,3,opt,name=readKeyIDHashes"`
```

Comment:
I thought the full set of key_ids required for decryption would be consistent for all kube-apiservers since it is logically the list of all known key_ids on the envelope.  If this is supposed to reflect the key_id that the KMS plugin can interpret, I don't remember an API to the kms plugin that gives us that list.